src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs

```diff
@@ -0,0 +1,824 @@
+            }
+        }
+
+        internal bool CanGetSecureStream
+        {
+            get
+            {
+                return (_context.IsConfidentialityFlag || _context.IsIntegrityFlag);
+            }
+        }
+
+        internal TokenImpersonationLevel AllowedImpersonation
+        {
+            get
+            {
+                CheckThrow(true);
+                return PrivateImpersonationLevel;
+            }
+        }
+
+        private TokenImpersonationLevel PrivateImpersonationLevel
+        {
+            get
+            {
+                // We should suppress the delegate flag in NTLM case.
+                return (_context.IsDelegationFlag && _context.ProtocolName != NegotiationInfoClass.NTLM) ? TokenImpersonationLevel.Delegation
+                        : _context.IsIdentifyFlag ? TokenImpersonationLevel.Identification
+                        : TokenImpersonationLevel.Impersonation;
+            }
+        }
+
+        private bool HandshakeComplete
+        {
+            get
+            {
+                return _context.IsCompleted && _context.IsValidContext;
+            }
+        }
+
+        internal void CheckThrow(bool authSucessCheck)
+        {
+            if (_exception != null)
+            {
+                throw _exception;
+            }
+
+            if (authSucessCheck && !IsAuthenticated)
+            {
+                throw new InvalidOperationException(SR.net_auth_noauth);
+            }
+        }
+
+        //
+        // This is to not depend on GC&SafeHandle class if the context is not needed anymore.
+        //
+        internal void Close()
+        {
+            // Mark this instance as disposed.
+            _exception = new ObjectDisposedException("NegotiateStream");
+            if (_context != null)
+            {
+                _context.CloseContext();
+            }
+        }
+
+        internal void ProcessAuthentication(LazyAsyncResult lazyResult)
+        {
+            CheckThrow(false);
+            if (Interlocked.Exchange(ref _nestedAuth, 1) == 1)
+            {
+                throw new InvalidOperationException(SR.Format(SR.net_io_invalidnestedcall, lazyResult == null ? "BeginAuthenticate" : "Authenticate", "authenticate"));
+            }
+
+            try
+            {
+                if (_context.IsServer)
+                {
+                    // Listen for a client blob.
+                    StartReceiveBlob(lazyResult);
+                }
+                else
+                {
+                    // Start with the first blob.
+                    StartSendBlob(null, lazyResult);
+                }
+            }
+            catch (Exception e)
+            {
+                // Round-trip it through SetException().
+                e = SetException(e);
+                throw;
+            }
+            finally
+            {
+                if (lazyResult == null || _exception != null)
```

Comment:
Doesn't this have the same bug that was fixed for SslStream at https://github.com/stephentoub/corefx/commit/9e63a418f6ab0dc01ab9e51f2ad9d498a35899cd ?